cctp_primitives/src/bit_vector/compression.rs

```diff
@@ -0,0 +1,276 @@
+use std::io::{Read, Write};
+
+use bzip2::read::{BzEncoder, BzDecoder};
+use flate2::{Compression as GzipCompression, write::GzEncoder, read::GzDecoder};
+
+use crate::{ printdbg, printlndbg};
+
+type Error = Box<dyn std::error::Error>;
+
+/// Available compression algorithms.
+/// The ffi repr(C) tag has been added here because this enum must be exported from mc-cryptolib.
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub enum CompressionAlgorithm {
+    Uncompressed,
+    Bzip2,
+    Gzip
+}
+
+impl TryFrom<u8> for CompressionAlgorithm {
+    type Error = ();
+
+    fn try_from(v: u8) -> Result<Self, Self::Error> {
+        match v {
+            x if x == CompressionAlgorithm::Uncompressed as u8 => Ok(CompressionAlgorithm::Uncompressed),
+            x if x == CompressionAlgorithm::Bzip2 as u8 => Ok(CompressionAlgorithm::Bzip2),
+            x if x == CompressionAlgorithm::Gzip as u8 => Ok(CompressionAlgorithm::Gzip),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Compresses `raw_bit vector` (represented as a byte vector)
+/// by using the specified compression `algorithm`.
+/// The resulting compressed bit vector has an additional first byte
+/// representing the algorithm used for the compression process.
+///
+/// # Examples
+///
+/// ```
+/// use cctp_primitives::bit_vector::compression::*;
+///
+/// let bit_vector: Vec<u8> = (0..100).collect();
+///
+/// let compressed_bit_vector = compress_bit_vector(&bit_vector, CompressionAlgorithm::Uncompressed).unwrap();
+/// assert_eq!(compressed_bit_vector[0], CompressionAlgorithm::Uncompressed as u8);
+/// assert_eq!(bit_vector.len() + 1, compressed_bit_vector.len());
+/// let decompressed_bit_vector = decompress_bit_vector(&compressed_bit_vector, bit_vector.len()).unwrap();
+/// assert_eq!(bit_vector, decompressed_bit_vector);
+///
+/// let compressed_bit_vector = compress_bit_vector(&bit_vector, CompressionAlgorithm::Bzip2).unwrap();
+/// assert_eq!(compressed_bit_vector[0], CompressionAlgorithm::Bzip2 as u8);
+/// let decompressed_bit_vector = decompress_bit_vector(&compressed_bit_vector, bit_vector.len()).unwrap();
+/// assert_eq!(bit_vector, decompressed_bit_vector);
+///
+/// let compressed_bit_vector = compress_bit_vector(&bit_vector, CompressionAlgorithm::Gzip).unwrap();
+/// assert_eq!(compressed_bit_vector[0], CompressionAlgorithm::Gzip as u8);
+/// let decompressed_bit_vector = decompress_bit_vector(&compressed_bit_vector, bit_vector.len()).unwrap();
+/// assert_eq!(bit_vector, decompressed_bit_vector);
+/// ```
+#[allow(unused_variables)]
+pub fn compress_bit_vector(raw_bit_vector: &[u8], algorithm: CompressionAlgorithm) -> Result<Vec<u8>, Error> {
+    let compressed_bit_vector_result;
+
+    printlndbg!("Compressing bit vector...");
+    printlndbg!("Algorithm: {}, size: {}, address: {:p}", algorithm as u8, raw_bit_vector.len(), raw_bit_vector);
+
+    printlndbg!("Bit vector content:");
+
+    raw_bit_vector.iter().for_each(|byte| printdbg!("|{}", byte));
```

Comment:
this can be expensive

Reply:
`printdbg` is a macro that has been specifically developed to work in _DEBUG_ mode only; when the code is compiled in RELEASE mode, it's not expanded, and then the compiler removes that code for optimization.

In any case, that print line has been refactored to be more concise and to make more evident that there is no performance degradation.